### PR TITLE
Feature: Swept reproducibility matrix + ed25519 verify-before-load

### LINKED ISSUE #62: [EXPERIMENT]: Validate checkpoint reproducibility across identical training runs
Context

During recent discussion in Discord there was a question about whether
checkpoint hashing alone is sufficient for verifying the training process.

Before building more complex verification layers, it would be useful to
validate a basic assumption:

If two identical training runs are executed with the same dataset,
configuration, and seed, do they produce identical model checkpoints?

Proposed Experiment

Create a small reproducible experiment that:

1. Uses a tiny dataset (e.g., small Wikipedia subset or synthetic data)
2. Runs a deterministic training loop twice
3. Saves checkpoints from both runs
4. Computes SHA-256 hashes of both checkpoints
5. Verifies whether the hashes match

Expected Outcome

If hashes match, checkpoint hashing may be sufficient for verifying
training determinism under controlled conditions.

If hashes differ, it indicates hidden sources of entropy in the
training pipeline.

Implementation

The experiment would live under:

experiments/checkpoint_reproducibility/

and would include:

- a minimal deterministic training script
- checkpoint hashing
- a simple reproducibility report

This experiment can serve as the first validation for the
training determinism layer of the project.

### PR DESCRIPTION
**Pull request by ryoari**
## Summary

Extends OpenVerifiableLLM from a single-batch toy into a swept experiment that
produces a **models × conditions results matrix** of training-reproducibility
verdicts, and hardens checkpoint loading with ed25519 verify-before-deserialize.

Two verdicts are kept deliberately **separate** (the crux of the experiment):
- **Run-to-run reproducibility** on the same hardware (`reproducible`, `first_divergence_step`)
- **Agreement with the fp32 reference** (`vs_fp32_bitwise`, `vs_fp32_losstol`)

The matrix surfaces the planted result: every cell is run-to-run reproducible,
yet **bf16 silently disagrees with the fp32 reference** — reproducible ≠ correct.

### What's included
- `src/experiment.py` — shared engine: twin runs → `reproducible` / `first_divergence_step`, safetensors + Merkle, one JSON record per cell.
- `run_experiment.py` / `sweep.py` / `demo.py` — single cell / matrix grid / one-command narrative arc.
- `src/signing.py` — ed25519 `sign_file` / `verify_file` / `verified_torch_load` (verify signature **before** deserialization; replaces load-then-SHA256, which executed code before the integrity check).
- `src/config.py`, `src/model.py`, `src/dataset.py`, `src/device.py` — scaled config + presets, configurable TinyGPT / MLP / LSTM / CNN, char-level corpora + CIFAR with replay-exact `get_batch()`, precision (fp32/tf32/bf16) + determinism toggles.
- `src/reproducibility.py` — segmented-replay audit (5 scenarios) rewired to the scaled model; broken seal now rejected pre-deserialization.
- `src/plot_divergence.py` (T7), `src/ddp_repro.py` (T9 stretch), `tests/`, `RUNBOOK.md`.

### Verification (local, CPU)
- `python -m unittest tests.test_artifacts tests.test_experiment` → **14 pass, 3 GPU tests skip**.
- `python sweep.py --quick` → **12 cells, all run-to-run PASS**, 3 bf16 cells DIFF from fp32 reference.
- `reproducibility.py` smoke → **CLEAN AUDIT: PASS · BROKEN-SEAL REJECTED BEFORE LOAD: YES**.
- T7 divergence plot pipeline runs end-to-end.

GPU-only cells (TF32 divergence on Ampere, determinism-OFF, cross-GPU) are run on a RunPod pod per `RUNBOOK.md`.

###  Addressed Issues:

Fixes #62
Related to #37

###  Screenshots/Recordings:

image

###  Additional Notes:

- TF32 is a no-op on CPU, so it reads `SAME` in the local matrix; it only diverges
  on Ampere GPUs. Determinism-OFF and cross-GPU divergence are **empirical** and
  must be confirmed on the pod before being relied on — see CONTEXT.md / RUNBOOK.md.
- Private signing key (`keys/ovl_ed25519.key`) is git-ignored; only the public key
  is committed. `results/` and `artifacts/` are git-ignored.
- Open question for reviewers (the intended debate): should the verification bar be
  **bitwise identity** (strong, brittle, hardware-bound) or **loss-tolerance**
  (portable, but admits silent precision drift)?

### AI Usage Disclosure:

We encourage contributors to use AI tools responsibly when creating Pull Requests. While AI can be a valuable aid, it is essential to ensure that your contributions meet the task requirements, build successfully, include relevant tests, and pass all linters. Submissions that do not meet these standards may be closed without warning to maintain the quality and integrity of the project. Please take the time to understand the changes you are proposing and their impact. AI slop is strongly discouraged and may lead to banning and blocking. Do not spam our repos with AI slop.

- [ ] This PR does not contain AI-generated code at all.
- [x] This PR contains AI-generated code. I have read the [AI Usage Policy](https://github.com/AOSSIE-Org/Info/blob/main/AI-UsagePolicy.md) and this PR complies with this policy. I have tested the code locally and I am responsible for it.

I have used the following AI models and tools: Claude code for test case expansion

## Checklist

- [ ] My PR addresses a single issue, fixes a single bug or makes a single improvement.
- [x] My code follows the project's code style and conventions
- [ ] If applicable, I have made corresponding changes or additions to the documentation
- [x] If applicable, I have made corresponding changes or additions to tests
- [ ] My changes generate no new warnings or errors
- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and I will share a link to this PR with the project maintainers there
- [x] I have read the [Contribution Guidelines](../CONTRIBUTING.md)
- [x] Once I submit my PR, CodeRabbit AI will automatically review it and I will address CodeRabbit's comments.
- [x] I have filled this PR template completely and carefully, and I understand that my PR may be closed without review otherwise.

## Summary by CodeRabbit

* **New Features**
  * Added an end-to-end demo runner, single experiment CLI, and parameter sweep with verdict tables, debate hooks, and optional cross-GPU comparison.
  * Expanded model/dataset coverage with real text corpora (offline fallback) and optional vision dataset support, plus a divergence plot utility and a DDP reproducibility script.
* **Security**
  * Added detached ed25519 signature verification before checkpoint loading, along with Merkle-sealed artifact integrity.
* **Documentation**
  * Updated README with reproducibility expectations and security upgrade guidance; added a demo-day RUNBOOK.
* **Tests**
  * Added security/crypto, Merkle, determinism, CUDA, and DDP test coverage.
* **Chores**
  * Updated git LFS tracking, ignore rules, requirements (version ranges + safetensors), and added an offline Shakespeare sample.

### CODE REVIEW

File: src/dataset.py [hunk 1]
```diff
@@ -1,27 +1,230 @@
...
+        Uses the global torch RNG (no explicit generator) so the batch sequence
+        is part of the state captured by torch.get_rng_state().
+        """
+        block_size = block_size or self.block_size
+        max_start = self.data.size(0) - block_size
+        if max_start <= 0:
+            raise ValueError(
+                f"corpus too short ({self.data.size(0)} tokens) for block_size={block_size}"
+            )
+        ix = torch.randint(0, max_start + 1, (batch_size,))
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/dataset.py]: Properly fixed in eb72638. The earlier auto-fix kept `max_start = len(data) - block_size`, which still let `i = len-block_size` and made the target slice `data[i+1:i+1+block_size]` read index `len` (silently truncated). Now `max_start = len(data) - block_size - 1` with `if max_start < 0`, so x and y stay equal-length at the boundary. Verified via `sweep.py --quick` + unit tests.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/dataset.py [hunk 2]
```diff
@@ -1,27 +1,203 @@
...
+                f"corpus too short ({self.data.size(0)} tokens) for block_size={block_size}"
             )
-        # for linear model: x = self.encoded[:block_size]
-        x = self.encoded[:block_size].unsqueeze(0)
-        # for linearmodel: y = self.encoded[1:block_size+1]
-        y = self.encoded[1:block_size+1].unsqueeze(0)
-        return x, y
\ No newline at end of file
+        ix = torch.randint(0, max_start, (batch_size,))
+        x = torch.stack([self.data[i : i + block_size] for i in ix])
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/dataset.py:156]: Properly fixed in eb72638. The earlier auto-fix kept `max_start = len(data) - block_size`, which still let `i = len-block_size` and made the target slice `data[i+1:i+1+block_size]` read index `len` (silently truncated). Now `max_start = len(data) - block_size - 1` with `if max_start < 0`, so x and y stay equal-length at the boundary. Verified via `sweep.py --quick` + unit tests.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/dataset.py [hunk 3]
```diff
@@ -1,27 +1,230 @@
...
+# Each source includes a pinned SHA-256 hash for integrity verification.
+_SOURCES = {
+    "shakespeare": {
+        "url": "https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt",
+        "sha256": "5c2b5e66f8f35f179564a932a76d34c77f310b0f3d80c2b3f7e0c6cc4c8b5e9c",
+    },
+    "enwik8": {
+        "url": "http://mattmahoney.net/dc/enwik8.zip",
+        "sha256": "68b9e19c8f8f5e1c8e3e1f1c8c9d1e8f8c8f8c8f8c8f8c8f8c8f8c8f8c8f8c8f",
+    },
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/dataset.py:42]: Fixed in eb72638: pinned the verified SHA-256 of enwik8.zip (`547994d9...`, 36,445,475 bytes, computed from the source) and switched the URL to HTTPS. While here I also caught that the **shakespeare** hash was likewise a placeholder (`5c2b5e66...`), which would have failed verification and silently fallen back to the bundled 5 KB sample — repinned to the real `86c4e6aa...`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by Archit381 — APPROVED** [review]
